ci/tests/smash_test.c

```diff
@@ -0,0 +1,41 @@
+/* Test that overwrite error detection works reasonably.        */
+
+#ifndef GC_DEBUG
+# define GC_DEBUG
+#endif
+
+#include "gc.h"
+
+#include <stdio.h>
+
+#define COUNT 7000
+#define SIZE  40
+
+char * A[COUNT];
+
+char * volatile q;
+
+int main(void)
+{
+  int i;
+  char *p;
+
+  GC_INIT();
+
+  for (i = 0; i < COUNT; ++i) {
+     A[i] = p = (char*)GC_MALLOC(SIZE);
+
+     if (i%3000 == 0) {
+        q = NULL;
+        GC_gcollect();
+     } else if (i%5678 == 0 && p != 0) {
+        /* Write a byte past the end of the allocated object    */
+        /* but not beyond the last word of the object's memory. */
+        /* A volatile intermediate pointer variable is used to  */
+        /* avoid a compiler complain of out-of-bounds access.   */
```

Comment:
TL;DR: This is probably fine, but the comment raises an eyebrow.

----

How does the compiler know that it's out of bounds? Is `GC_MALLOC` annotated with some special `attribute`?

Using `volatile` just to disable a warning is probably not ideal — there'll be a `-Wno-...` option for that — but it does make it less likely that the compiler will discard the whole access as undefined behaviour. (It's still undefined behaviour, but `volatile` tends not to be as aggressively optimised so we'll probably get away with it.)

Reply:
I agree that this test looks pretty weird (and I also had the same wide-eyed stare at the "volatile" bit), but since it's already part of the test suite, we might as well take it as a given.

Reply:
Agreed!